README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# ngboostForecast

<!-- badges: start -->
<!-- badges: end -->

The goal of ngboostForecast is to provide a tools for probabilistic
forecasting by using Python’s ngboost for R users.

## Installation

You can install the released version of ngboostForecast from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("ngboostForecast")
```

And the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("Akai01/ngboostForecast")
```

## Example

This is a basic example which shows you how to solve a common problem:

``` r
library(ngboostForecast)
#> Loading required package: reticulate
#> Registered S3 method overwritten by 'quantmod':
#>   method            from
#>   as.zoo.data.frame zoo

train = window(seatbelts, end = c(1983,12))

test = window(seatbelts, start = c(1984,1))

# without external variables with Ridge regression

model <- NGBforecast$new(Dist = Dist("LogNormal"),
                         Base = sklearner(module = "linear_model",
                                          class = "Ridge"),
                         Score = Scores("LogScore"),
                         natural_gradient = TRUE,
                         n_estimators = 200,
                         learning_rate =  0.1,
                         minibatch_frac = 1,
                         col_sample = 1,
                         verbose = TRUE,
                         verbose_eval = 5,
                         tol = 1e-5)

model$fit(y = train[,2], 
          seasonal = TRUE, 
          max_lag = 12, 
          early_stopping_rounds = 10L)

fc <- model$forecast(h = 12, level = c(99,95,90, 80, 70, 60), 
                      data_frame = FALSE)

autoplot(fc) + autolayer(test[,2])
```

<img src="man/figures/README-example-1.png" width="100%" />

# Tuning

## Set the parameters:

``` r
library(ngboostForecast)

dists <- list(Dist("Normal"))

base_learners <- list(sklearner(module = "tree", class = "DecisionTreeRegressor",
                                max_depth = 1),
                      sklearner(module = "tree", class = "DecisionTreeRegressor",
                                max_depth = 2),
                      sklearner(module = "tree", class = "DecisionTreeRegressor",
                                max_depth = 3),
                      sklearner(module = "tree", class = "DecisionTreeRegressor",
                                max_depth = 4),
                      sklearner(module = "tree", class = "DecisionTreeRegressor",
                                max_depth = 5),
                      sklearner(module = "tree", class = "DecisionTreeRegressor",
                                max_depth = 6),
                      sklearner(module = "tree", class = "DecisionTreeRegressor",
                                max_depth = 7))

scores <-  list(Scores("LogScore"))

model <- NGBforecastCV$new(Dist = dists,
                           Base = base_learners,
                           Score = scores,
                           natural_gradient = TRUE,
                           n_estimators = list(10, 100),
                           learning_rate = list(0.1, 0.2),
                           minibatch_frac = list(0.1, 1),
                           col_sample = list(0.3),
                           verbose = FALSE,
                           verbose_eval = 100,
                           tol = 1e-5)
```

## Tune the model:

``` r
params <- model$tune(y = AirPassengers,
                     seasonal = TRUE,
                     max_lag = 12,
                     xreg = NULL,
                     early_stopping_rounds = NULL,
                     n_splits = 4L)
```

## Best parameters:

``` r
params
#> $ngboost_best_params
#> $ngboost_best_params$Base
#> DecisionTreeRegressor(max_depth=7.0)
#> 
#> $ngboost_best_params$Dist
#> <class 'ngboost.distns.normal.Normal'>
#> 
#> $ngboost_best_params$Score
#> <class 'ngboost.scores.LogScore'>
#> 
#> $ngboost_best_params$col_sample
#> [1] 0.3
#> 
#> $ngboost_best_params$learning_rate
#> [1] 0.2
#> 
#> $ngboost_best_params$minibatch_frac
#> [1] 1
#> 
#> $ngboost_best_params$n_estimators
#> [1] 100
#> 
#> 
#> $ngb_forecast_params
#> $ngb_forecast_params$seasonal
#> [1] TRUE
#> 
#> $ngb_forecast_params$max_lag
#> [1] 12
#> 
#> $ngb_forecast_params$K
#> [1] 5
```


NEWS.md
# ngboostForecast 0.1.1

# ngboostForecast 0.1.0

# ngboostForecast 0.0.2

# ngboostForecast 0.0.1

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(ngboostForecast)

if_not_ngboost_exist_skip <- function() {
  clist <- FALSE
  clist <- reticulate::py_module_available("ngboost")

  if (!clist) {
    testthat::skip("ngboost not available for testing")
  }
}

test_check("ngboostForecast")


tests/testthat/test-ngboostfor.R
# A unit test for NGBforecast class
if(require(testthat)){
  test_that("tests for some arguments in NGBforecast", {

    if_not_ngboost_exist_skip()

    model <- NGBforecast$new(Dist = Dist("LogNormal"),
                             Base = sklearner(),
                             Score = Scores("LogScore"),
                             natural_gradient = TRUE,
                             n_estimators = 200,
                             learning_rate =  0.1,
                             minibatch_frac = 1,
                             col_sample = 1,
                             verbose = TRUE,
                             verbose_eval = 100,
                             tol = 1e-5)

    model$fit(y = AirPassengers, seasonal = TRUE, max_lag = 12, xreg = NULL,
              early_stopping_rounds = 10L, K = 5)


    fc <- c(model$forecast(h = 2, level = c(90, 80), xreg = NULL)$mean)

    expect_equal(fc, c(454.3927, 425.1821), tolerance = 4)

  })
  
  test_that("tests for non-seasonal NGBforecast", {
    
    if_not_ngboost_exist_skip()
    
    model <- NGBforecast$new(Dist = Dist("LogNormal"),
                             Base = sklearner(),
                             Score = Scores("LogScore"),
                             natural_gradient = TRUE,
                             n_estimators = 200,
                             learning_rate =  0.1,
                             minibatch_frac = 1,
                             col_sample = 1,
                             verbose = TRUE,
                             verbose_eval = 100,
                             tol = 1e-5)
    
    model$fit(y = AirPassengers, seasonal = FALSE, max_lag = 12, xreg = NULL, 
              early_stopping_rounds = 10L)
    
    
    fc <- c(model$forecast(h = 2, level = c(90, 80), xreg = NULL)$mean)
    
    expect_equal(fc, c(468, 461), tolerance = 5)
    
  })
  
  test_that("tests for non-seasonal xreg NGBforecast", {
    
    if_not_ngboost_exist_skip()
    
    model <- NGBforecast$new(Dist = Dist("LogNormal"),
                             Base = sklearner(),
                             Score = Scores("LogScore"),
                             natural_gradient = TRUE,
                             n_estimators = 200,
                             learning_rate =  0.1,
                             minibatch_frac = 1,
                             col_sample = 1,
                             verbose = TRUE,
                             verbose_eval = 100,
                             tol = 1e-5)
    
    xreg <- matrix(rnorm(length(AirPassengers)*2, mean = 3), ncol = 2)
    newxreg <- matrix(rnorm(4, mean = 3), ncol = 2)
    
    model$fit(y = AirPassengers, seasonal = FALSE, max_lag = 12, xreg = xreg, 
              early_stopping_rounds = 10L)
    
    
    fc <- c(model$forecast(h = 2, level = c(90, 80), xreg = newxreg)$mean)
    
    expect_equal(fc, c(463, 461), tolerance = 5)
    
  })
  
  test_that("tests for seasonal xreg NGBforecast", {
    
    if_not_ngboost_exist_skip()
    
    model <- NGBforecast$new(Dist = Dist("LogNormal"),
                             Base = sklearner(),
                             Score = Scores("LogScore"),
                             natural_gradient = TRUE,
                             n_estimators = 200,
                             learning_rate =  0.1,
                             minibatch_frac = 1,
                             col_sample = 1,
                             verbose = TRUE,
                             verbose_eval = 100,
                             tol = 1e-5)
    
    xreg <- matrix(rnorm(length(AirPassengers)*2, mean = 3), ncol = 2)
    newxreg <- matrix(rnorm(4, mean = 3), ncol = 2)
    
    model$fit(y = AirPassengers, seasonal = TRUE, max_lag = 12, xreg = xreg, 
              early_stopping_rounds = 10L)
    
    
    fc <- c(model$forecast(h = 2, level = c(90, 80), xreg = newxreg)$mean)
    
    expect_equal(fc, c(470, 463), tolerance = 5)
    
  })
  
}


tests/testthat/test-ngbforecastcv.R
if(require(testthat)){
  test_that("tests for some arguments in NGBforecastCV", {
    
    if_not_ngboost_exist_skip()
    
    dists <- list(Dist("Normal"))
    base_learners <- list(sklearner(module = "tree", class = "DecisionTreeRegressor",
                                    max_depth = 6),
                          sklearner(module = "tree", class = "DecisionTreeRegressor",
                                    max_depth = 7))
    scores <-  list(Scores("LogScore"))
    
    model <- NGBforecastCV$new(Dist = dists,
                               Base = base_learners,
                               Score = scores,
                               natural_gradient = TRUE,
                               n_estimators = list(10, 12),
                               learning_rate = list(0.1),
                               minibatch_frac = list(0.1),
                               col_sample = list(0.3),
                               verbose = FALSE,
                               verbose_eval = 100,
                               tol = 1e-5)
    
    params <- model$tune(y = AirPassengers,
                         seasonal = TRUE,
                         max_lag = 12,
                         xreg = NULL,
                         early_stopping_rounds = NULL,
                         n_splits = 4L)
    
    out <- class(params)
    
    expect_equal(out, "list")
    
  })
  
}

R/utils.R
#' Scikit-Learn interface
#' @param module scikit-learn module name, default is 'tree'.
#' @param class scikit-learn's module class, default is 'DecisionTreeRegressor'
#' @param ... Other arguments passed to model class
#' @author Resul Akay
#' @examples
#' \dontrun{
#'
#' sklearner(module = "tree", class = "DecisionTreeRegressor",
#' criterion="friedman_mse", min_samples_split=2)
#'
#' }
#'
#' @export
sklearner <- function(module = "tree", class = "DecisionTreeRegressor",
                      ...){
  sklearn[[module]][[class]](...)
}

#' NGBoost distributions
#' @param dist NGBoost distributions. One of the following:
#' \itemize{
#' \item Bernoulli
#' \item k_categorical
#' \item StudentT
#' \item Poisson
#' \item Laplace
#' \item Cauchy
#' \item Exponential
#' \item LogNormal
#' \item MultivariateNormal
#' \item Normal
#' }
#' @import reticulate
#' @param k Used only with k_categorical and MultivariateNormal
#' @returns An NGBoost Distribution object
#' @export
Dist <- function(dist = c("Normal", "Bernoulli", "k_categorical", "StudentT",
                          "Laplace", "Cauchy", "Exponential", "LogNormal",
                          "MultivariateNormal", "Poisson"), k){
  dist <- match.arg(dist)
  if(dist %in% c("MultivariateNormal", "k_categorical")){
    if(dist == "k_categorical"){
      out <- ngboost[["distns"]][[dist]](K = k)
    } else{
      out <- ngboost[["distns"]][[dist]](k = k)
    }
  } else {
    out <- ngboost[["distns"]][[dist]]
  }
  out
}

#' Select a rule to compare probabilistic predictions to the observed data.
#' @description 
#' Select a rule to compare probabilistic predictions to the observed data.
#' A score from ngboost.scores, e.g. LogScore.
#' @param score A string. can be one of the following:
#' \itemize{
#'
#'  \item LogScore : Generic class for the log scoring rule.
#'  \item CRPS : Generic class for the continuous ranked probability scoring rule.
#'  \item CRPScore : Generic class for the continuous ranked probability scoring rule.
#'  \item MLE : Generic class for the log scoring rule.
#'  }
#' @author Resul Akay
#' @return A score class from ngboost.scores
#' @export
Scores <- function(score = c("LogScore", "CRPS", "CRPScore", "MLE")){
  score <- match.arg(score)
  out <- scores[[score]]
  return(out)
}

#' Is conda installed?
#' @description 
#' Only for internal usage. 
#' @return Logical, TRUE if conda is installed.
#' @export
#' @author Resul Akay
is_exists_conda <- function() {
  clist <- tryCatch({
    reticulate::conda_list()
  }, error = function(e){
    data.frame()
  }
  )
  return(length(clist) > 0)
}

#' Pipe operator
#'
#' See \code{magrittr::\link[magrittr:pipe]{\%>\%}} for details.
#'
#' @name %>%
#' @rdname pipe
#' @keywords internal
#' @export
#' @importFrom magrittr %>%
#' @usage lhs \%>\% rhs
NULL


#' Road Casualties in Great Britain 1969-84
#'
#' @description
#' The Seatbelts dataset from the datasets package.
#' @source
#' Harvey, A.C. (1989). Forecasting, Structural Time Series Models and the
#'  Kalman Filter. Cambridge University Press, pp. 519–523.
#'
#'
#' Durbin, J. and Koopman, S. J. (2001). Time Series Analysis by State Space
#' Methods. Oxford University Press.
#'
#'
#' \url{https://stat.ethz.ch/R-manual/R-devel/library/datasets/html/UKDriverDeaths.html}
#'
#' @references
#' Harvey, A. C. and Durbin, J. (1986). The effects of seat belt legislation on
#' British road casualties: A case study in structural time series modelling.
#' Journal of the Royal Statistical Society series A, 149, 187–227.
#'
"seatbelts"



#' Assignment pipe operator
#'
#' See \code{magrittr::\link[magrittr:pipe]{\%<>\%}} for details.
#'
#' @name %<>%
#' @rdname assignment_pipe
#' @keywords internal
#' @export
#' @importFrom magrittr %<>%
#' @usage lhs \%<>\% rhs
NULL

#' forecast package autoplot function
#'
#' See \code{\link[forecast]{autoplot}} for details.
#'
#' @name autoplot
#' @rdname autoplot
#' @keywords internal
#' @export
#' @importFrom forecast autoplot
#' @usage autoplot(object,...)
#' @return A ggplot object
#' @seealso \code{\link[forecast]{autoplot}}
NULL

#' forecast package autolayer function
#'
#' See \code{\link[forecast]{autolayer}} for details.
#'
#' @name autolayer
#' @rdname autolayer
#' @keywords internal
#' @export
#' @importFrom forecast autolayer
#' @return A ggplot layer
#' @usage autolayer(object,...)
#' @seealso \code{\link[forecast]{autolayer}}
NULL

#' magrittr pipe operator
#'
#' See \code{magrittr::\link[magrittr:pipe]{\%>\%}} for details.
#'
#' @name %>%
#' @rdname pipe
#' @keywords internal
#' @export
#' @importFrom magrittr %>%
#' @return Nothing
#'
#' @usage lhs \%>\% rhs
NULL


`%notin%` <- Negate(`%in%`)

#' @importFrom stats na.omit
lag_maker <- function(y, max_lag) {
  if ("ts" %notin% class(y)) {
    stop("y must be a 'ts' object")
  }
  max_lag1 <- round(max_lag)
  if (max_lag1 != max_lag) {
    message(paste("'max_lag' should not be a fractional number.", 
                  "'max_lag' rounde to",  max_lag1, sep = " "))
  }
  y <- c(y)
  out <- matrix(nrow = length(y), ncol = max_lag)
  for (i in seq_len(max_lag)) {
    out[, i] <- dplyr::lag(y, i)
  }
  out <- as.matrix(na.omit(out))
  colnames(out) <- paste0("y_lag", seq_len(max_lag))
  return(out)
}

#' @importFrom stats frequency time ts
prepare_data <- function(y,
                         max_lag = 5,
                         xreg = NULL,
                         seasonal = TRUE,
                         K =  frequency(y) / 2 - 1){

  if ("ts" %notin% class(y)) {
    stop("y must be a univariate time series")
  }
  
  length_y <- length(y)
  freq <- stats::frequency(y)
  
  if (length_y < freq) {
    stop("Not enough data to fit a model")
  }
  if (max_lag <= 0) {
    warning("max_lag increased to 1. max_lag must be max_lag >= 1")
    max_lag <- 1
  }
  if (c(length_y - freq - round(freq / 4)) < max_lag) {
    warning(paste("Input data is too short. Reducing max_lags to ",
                  round(length_y - freq - round(freq / 4))))
    max_lag <- round(length_y - freq - round(freq / 4))
  }
  if (max_lag != round(max_lag)) {
    max_lag <- round(max_lag)
    message(paste("max_lag must be an integer, max_lag rounded to", max_lag))
  }
  if (!is.null(xreg)) {
    if ("matrix" %notin% class(xreg)) {
      xreg <- as.matrix(xreg)
    }
  }
  constant_y <- forecast::is.constant(forecast::na.interp(y))
  if (constant_y) {
    warning("Constant data, setting max_lag = 1")
    max_lag = 1
  }

  if (!is.null(xreg))
  {
    ncolxreg <- ncol(xreg)
  }

  modified_y <-
    ts(y[-seq_len(max_lag)], start = time(y)[max_lag + 1], frequency = freq)
  
  if (seasonal == TRUE | freq > 1)
  {
    if (K == freq / 2) {
      ncolx <- max_lag + K * 2 - 1
    } else {
      ncolx <- max_lag + K * 2
    }
  }
  if (seasonal == FALSE | freq == 1)
  {
    ncolx <- max_lag
  }

  x <- matrix(0, nrow = c(length_y - max_lag), ncol = ncolx)

  x[, seq_len(max_lag)] <- lag_maker(y, max_lag)
  
  fourier_s <- NULL
  
  if (seasonal == TRUE & freq > 1)
  {
    fourier_s <- forecast::fourier(modified_y, K = K)
    x[, (max_lag + 1):ncolx] <- fourier_s
    colnames(x) <- c(paste0("lag", 1:max_lag), colnames(fourier_s))
  }

  if (seasonal == FALSE | freq == 1)
  {
    colnames(x) <- c(paste0("lag", 1:max_lag))
  }

  if (!is.null(xreg)) {
    col_xreg <- ncol(xreg)
    name_xreg <- colnames(xreg)
    xreg <- xreg[-seq_len(max_lag), ]
    if (col_xreg == 1) {
      xreg <- as.matrix(matrix(xreg, ncol = 1))
      colnames(xreg)[1] <- name_xreg[1]
      rm(name_xreg, col_xreg)
    }
    x <- cbind(x, xreg)
  }
  return(list("x" = x,
              "y" = modified_y,
              "frequency" = freq,
              "fourier_s" = fourier_s))
}


R/package.R
#' @title Probabilistic Time Series Forecasting
#' 
#' @description Probabilistic time series forecasting via Natural Gradient 
#' Boosting for Probabilistic Prediction.
#' 
#' @rdname ngboostForecast
#' @docType package
#' @name ngboostForecast
#' @references 
#' 
#' Duan, T et. al. (2019), NGBoost: Natural Gradient Boosting for Probabilistic 
#' Prediction. 
#' @examples  
#' 
#' \dontrun{
#' 
#' library(ngboostForecast)
#' 
#' model <- NGBforecast$new(Dist = Dist("Normal"),
#'                          Base = sklearner(module = "linear_model",
#'                          class = "Ridge"),
#'                          Score = Scores("LogScore"),
#'                          natural_gradient = TRUE,
#'                          n_estimators = 200,
#'                          learning_rate =  0.1,
#'                          minibatch_frac = 1,
#'                          col_sample = 1,
#'                          verbose = TRUE,
#'                          verbose_eval = 100,
#'                          tol = 1e-5)
#' model$fit(y = AirPassengers, seasonal = TRUE, max_lag = 12, xreg = NULL,
#' 
#' early_stopping_rounds = 10L)
#' 
#' fc <- model$forecast(h = 12, level = c(90, 80), xreg = NULL)
#' 
#' autoplot(fc)
#' 
#'}
NULL


ngboost <- NULL
sklearn <- NULL
scores <- NULL

#' @export
.onLoad <- function(libname, pkgname) {
  reticulate::configure_environment(pkgname)
  ngboost <<- reticulate::import("ngboost",delay_load = TRUE)
  sklearn <<- reticulate::import("sklearn",delay_load = TRUE)
  scores <<- reticulate::import("ngboost.scores",delay_load = TRUE)
}



R/ngbforcv.R
#' @title NGBoost forecasting model selection class
#' @description
#' It is a wrapper for the sklearn GridSearchCV with TimeSeriesSplit.
#'
#' @examples
#' \dontrun{
#'
#' library(ngboostForecast)
#'
#' dists <- list(Dist("Normal"))
#'
#' base_learners <- list(sklearner(module = "tree", class = "DecisionTreeRegressor",
#'                                 max_depth = 1),
#'                       sklearner(module = "tree", class = "DecisionTreeRegressor",
#'                                 max_depth = 2),
#'                       sklearner(module = "tree", class = "DecisionTreeRegressor",
#'                                 max_depth = 3),
#'                       sklearner(module = "tree", class = "DecisionTreeRegressor",
#'                                 max_depth = 4),
#'                       sklearner(module = "tree", class = "DecisionTreeRegressor",
#'                                 max_depth = 5),
#'                       sklearner(module = "tree", class = "DecisionTreeRegressor",
#'                                 max_depth = 6),
#'                       sklearner(module = "tree", class = "DecisionTreeRegressor",
#'                                 max_depth = 7))
#'
#' scores <-  list(Scores("LogScore"))
#'
#' model <- NGBforecastCV$new(Dist = dists,
#'                            Base = base_learners,
#'                            Score = scores,
#'                            natural_gradient = TRUE,
#'                            n_estimators = list(10, 100),
#'                            learning_rate = list(0.1, 0.2),
#'                            minibatch_frac = list(0.1, 1),
#'                            col_sample = list(0.3),
#'                            verbose = FALSE,
#'                            verbose_eval = 100,
#'                            tol = 1e-5)
#'
#' params <- model$tune(y = AirPassengers,
#' seasonal = TRUE,
#' max_lag = 12,
#' xreg = NULL,
#' early_stopping_rounds = NULL,
#' n_splits = 4L)
#'
#' params
#'
#'}
#' @references
#' \url{https://stanfordmlgroup.github.io/ngboost/2-tuning.html}
#'
#' @author Resul Akay
#' @importFrom R6 R6Class
#' @export
NGBforecastCV <- R6::R6Class(
  classname = "NGBforecastCV",
  public = list(
    #' @description Initialize an NGBforecastCV model.
    #' @param Dist Assumed distributional form of \code{Y|X=x}. An output of 
    #' \code{\link{Dist}} function, e.g. \code{Dist('Normal')}
    #' 
    #' @param Score Rule to compare probabilistic predictions to 
    #' the observed data. A score from \code{\link{Scores}} function, e.g. 
    #' \code{Scores(score = "LogScore")}.
    #' @param Base Base learner. An output of \code{\link{sklearner}} function,
    #' e.g. \code{sklearner(module = "tree", class = "DecisionTreeRegressor", ...)}
    #' @param natural_gradient Logical flag indicating whether the natural 
    #' gradient should be used
    #' @param n_estimators The number of boosting iterations to fit
    #' @param learning_rate The learning rate
    #' @param minibatch_frac The percent subsample of rows to use in each 
    #' boosting iteration
    #' @param col_sample The percent subsample of columns to use in each 
    #' boosting iteration
    #' @param verbose Flag indicating whether output should be printed 
    #' during fitting. If TRUE it will print logs.
    #' @param verbose_eval Increment (in boosting iterations) at which 
    #' output should be printed
    #' @param tol Numerical tolerance to be used in optimization
    #' @param random_state Seed for reproducibility.
    #' @return An NGBforecastCV object that can be fit.
    initialize = function(Dist = NULL,
                          Score = NULL,
                          Base = NULL,
                          natural_gradient = TRUE,
                          n_estimators = as.integer(500),
                          learning_rate = 0.01,
                          minibatch_frac = 1.0,
                          col_sample = 1.0,
                          verbose = TRUE,
                          verbose_eval = as.integer(100),
                          tol = 0.0001,
                          random_state = NULL){

      if(!inherits(Dist, "list")){
        stop(
          "Please profide a list of Dist object with at least one spesification"
          )
      }
      if(!inherits(Score, "list")){
        stop(
          "Please profide a list of Score object with at least one spesification"
        )
      }
      if(!inherits(Base, "list")){
        stop(
          "Please profide a list of Base base learnesr with at least one spesification"
        )
      }
      if(!inherits(n_estimators, "list")){
        stop(
          "Please profide a list of n_estimators object with at least one spesification"
        )
      }
      if(!inherits(learning_rate, "list")){
        stop(
          "Please profide a list of learning_rate object with at least one spesification"
        )
      }
      if(!inherits(minibatch_frac, "list")){
        stop(
          "Please profide a list of minibatch_frac object with at least one spesification"
        )
      }
      if(!inherits(col_sample, "list")){
        stop(
          "Please profide a list of col_sample object with at least one spesification"
        )
      }

      private$Dist <- Dist
      private$Base <- Base
      private$Score <- Score
      private$natural_gradient <- natural_gradient
      private$n_estimators <- as.integer(n_estimators)
      private$learning_rate <- learning_rate
      private$minibatch_frac <- minibatch_frac
      private$col_sample <- col_sample
      private$verbose <- verbose
      private$verbose_eval <- as.integer(verbose_eval)
      private$tol <- tol
      private$random_state <- random_state

      if(!ngboostForecast::is_exists_conda()) {
        stop(
          paste(
            "\nConda is not available!",
            "\nPlease use",
            "'reticulate::install_miniconda(update = TRUE, force = TRUE)'",
            "and then restart R.'"
          )
        )
      }

      private$model <- ngboost$NGBRegressor(
        natural_gradient = private$natural_gradient,
        verbose = private$verbose,
        verbose_eval = private$verbose_eval,
        tol = private$tol,
        random_state = private$random_state
      )
      return(self)
    },
    #' @description Tune ngboosForecast.
    #' 
    #' @param y A time series (ts) object
    #' @param max_lag Maximum number of lags
    #' @param xreg Optional. A numerical matrix of external regressors,
    #' which must have the same number of rows as y. 
    #' @param test_size The length of validation set. 
    #' If it is NULL, then, it is automatically specified.
    #' @param seasonal Boolean. If \code{seasonal = TRUE} the fourier terms 
    #' will be used for modeling seasonality.
    #' @param K Maximum order(s) of Fourier terms, used only if
    #' \code{seasonal = TRUE}.
    #' @param n_splits Number of splits. Must be at least 2.
    #' @param train_loss_monitor A custom score or set of scores to track on the
    #' training set during training. Defaults to the score defined in the NGBoost
    #' constructor. Please do not modify unless you know what you are doing.
    #' @param val_loss_monitor A custom score or set of scores to track on the
    #'  validation set during training. Defaults to the score defined in the
    #'  NGBoost  constructor. Please do not modify unless you know what you are
    #'  doing.
    #' @param early_stopping_rounds The number of consecutive boosting
    #'  iterations during which the loss has to increase before the algorithm
    #'  stops early.
    #'  
    #'  
    #' @return A named list of best parameters.
    #'
    #' @importFrom forecast is.constant na.interp fourier
    tune = function(y,
                    max_lag = 5,
                    xreg = NULL,
                    seasonal = TRUE,
                    K =  frequency(y) / 2 - 1,
                    n_splits = NULL,
                    train_loss_monitor=NULL,
                    val_loss_monitor=NULL,
                    early_stopping_rounds=NULL){

      prepared_d <- prepare_data(y = y,
                                 max_lag = max_lag,
                                 xreg = xreg,
                                 seasonal = seasonal,
                                 K = K)

      X = prepared_d$x
      Y = prepared_d$y

      private$feature_names <- colnames(X)
      model = private$model

      param_grid = list("n_estimators" = private$n_estimators,
                        "learning_rate" = private$learning_rate,
                        "minibatch_frac" = private$minibatch_frac,
                        "col_sample" = private$col_sample,
                        "Base" = private$Base,
                        "Dist" = private$Dist,
                        "Score" = private$Score)

      tscv = sklearn[["model_selection"]]$TimeSeriesSplit(
        n_splits = as.integer(n_splits)
        )

      grdsrch<- sklearn[["model_selection"]]$GridSearchCV(
        model, param_grid=param_grid, cv = tscv
        )
      result <- grdsrch$fit(X,
                            Y,
                            train_loss_monitor = train_loss_monitor,
                            val_loss_monitor = val_loss_monitor,
                            early_stopping_rounds = early_stopping_rounds)

      out <- list(
        "ngboost_best_params" = result$best_params_,
        "ngb_forecast_params" = list(
          "seasonal" = seasonal,
          "max_lag" = max_lag,
          "K" = K
        )
      )

      return(out)
    }
  ),
  private = list(Dist = NULL,
                 Score = NULL,
                 Base = NULL,
                 natural_gradient = NULL,
                 n_estimators = NULL,
                 learning_rate = NULL,
                 minibatch_frac = NULL,
                 col_sample = NULL,
                 verbose = NULL,
                 verbose_eval = NULL,
                 tol = NULL,
                 random_state = NULL,
                 model = NULL,
                 feature_names = NULL)
)
